core/apple/src/AppleCore.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.NSInputStream
+import platform.Foundation.NSOutputStream
+import platform.Foundation.NSStreamStatusClosed
+import platform.Foundation.NSStreamStatusNotOpen
+import platform.darwin.UInt8Var
+
+/**
+ * Returns [RawSink] that writes to an output stream.
+ *
+ * Use [RawSink.buffered] to create a buffered sink from it.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.outputStreamAsSink
+ */
+public fun NSOutputStream.asSink(): RawSink = OutputStreamSink(this)
+
+private open class OutputStreamSink(
+    private val out: NSOutputStream,
+) : RawSink {
+
+    @OptIn(UnsafeNumber::class)
+    override fun write(source: Buffer, byteCount: Long) {
+        if (out.streamStatus == NSStreamStatusClosed) throw IOException("Stream Closed")
```

Comment:
Should we check other possible status values here?
For example, check if it's `NSStreamStatusWriting` and throw an exception in that case.

Reply:
Other status errors should cause an exception to be thrown on [this line](https://github.com/Kotlin/kotlinx-io/pull/174/files#diff-5bb5a6a76d27b1ed60b0526e03021a31b1b429883dd0017726c83edacf44347aR47). This handles `NSStreamStatusError` with an error message from the underlying `NSError`. However, a concurrency error would likely result in `Unknown error`, since it's not a terminal error.

Since we're not providing thread safety guarantees, checking for `NSStreamStatusWriting` here might give the illusion of thread safety. A successful check at the beginning of the function could still result in a failed `write()` call, if the other thread hasn't begun the `write()` operation until right after the check. Similarly a failed check would have no guarantee the `write()` call wouldn't have succeeded, had the other thread completed before we actually got to the call. And in either case, it's likely there'd be buffer state corruption by updating its state concurrently.